### ISSUE TITLE
Package request: Docker engine

### ISSUE THREAD

**Issue by danielmeneses** [termux/termux-packages#320]
Great job for all that Termux is. I know that this is probably a "no can do", but would be a for a nobel award!

;)

**Comment by samos123** [comment]
+1 this would also make it easier to support building termux packages on Android via docker. I guess the x86 architecture makes most sense, although ARM is also supported: https://github.com/umiddelb/armhf/wiki/Installing,-running,-using-docker-on-armhf-(ARMv7)-devices

**Comment by AleffSouza** [comment]
+1

**Comment by casept** [comment]
I have tried building docker engine on android. The major issue seems to be the craptastic, outdated kernels shipped with most devices, missing support for storage drivers in these kernels and the need for root to use docker.

**Comment by Serkan-devel** [comment]
Maybe it would work on lineage os

**Comment by casept** [comment]
It might on *some* extremely new devices. Most still package 3.x series kernels, which are too old for docker and can't be updated by ROM maintainers because most are raging tire fires of hackish vendor patches stacked one onto another.

**Comment by phonicmouse** [comment]
+1

**Comment by T-vK** [comment]
+1

**Comment by wrsg** [comment]
My device (Nexus 6) has a 3.10 kernel, I think it would be enough for Docker.

**Comment by T-vK** [comment]
Requirements seem to be:

- A 64-bit installation
- Version 3.10 or higher of the Linux kernel. The latest version of the kernel available for you platform is recommended.
- `iptables` version 1.4 or higher
- `git` version 1.7 or higher
- A `ps` executable, usually provided by `procps` or a similar package.
- `XZ Utils` 4.9 or higher
- a properly mounted `cgroupfs` hierarchy; a single, all-encompassing `cgroup` mount point is not sufficient.

**Comment by wrsg** [comment]
Damn, 64-bit only. That will rule it out for a lot of people.

**Comment by TheDiamondYT1** [comment]
I'm sure docker could be patched to support 32 bit, I could be wrong

**Comment by Serkan-devel** [comment]
I have docker running on my 32-bit raspberry pi 2

**Comment by T-vK** [comment]
@casept  I just noticed my 2014 tablet has a 3.10 kernel. So I think the number of devices that would support Docker might have risen quite a bit. Also, I don't think root is required. On the desktop you can simply add a non-root user to the docker group and then that user can run docker without root privileges.

**Comment by danielmeneses** [comment]
@T-vK that's nice and I don't quite know the docker implementation but if it makes use of the chroot command or similar I think we will need root access anyway. But it's somehow possible it would be great.

**Comment by wrsg** [comment]
https://docs.docker.com/engine/installation/linux/docker-ce/debian/
https://docs.docker.com/engine/installation/linux/docker-ce/ubuntu/

Debian and Ubuntu both have armhf Docker packages. And I don't know if there's anything to be gained from looking at the Moby project but it looks like this is where Docker system integration development is taking place now. 

https://github.com/moby/moby
https://mobyproject.org/

Hoping to get some time to look into this soon.

**Comment by smiller171** [comment]
@T-vK the docker client can run without root, but the docker engine requires root access afaik.

Also, it doesn't just need a 3.10 kernel, it needs kernel modules that aren't built into Android kernels. You would need to compile a custom kernel

**Comment by pwdonald** [comment]
I see that CrossOver runs Windows apps using emulation, is something similar possible with Linux on Android so that a more appropriate Kernel version could be utilized to run Docker?

**Comment by smiller171** [comment]
@pwdonald Nope. You either need to compile a custom Android kernel, or somehow run a VM on Android. Would be better just to run a remote server and SSH in.

**Comment by sneak** [comment]
The engine and all its associated architecture requirements aside, I would love to see just the docker client packaged.  One can use it with `socat` to redirect a local socket via an SSH tunnel to a remote host that is running the docker engine, or with the docker TCP/TLS rpc system (e.g. as set up by `docker-machine`).

**Comment by onlurking** [comment]
I don't think this is easy since Docker uses a [lot of the linux kernel features](https://medium.com/@nagarwal/understanding-the-docker-internals-7ccb052ce9fe), and Android ships a stripped down version of the kernel.

**Comment by smiller171** [comment]
The Docker client can be done. There are already Android apps that can talk to a remote Docker Engine if you expose the API

**Comment by sdrausty** [comment]
Initially this looks really promising through [Arch Linux in Termux PRoot](https://sdrausty.github.io/TermuxArch/):
```
pacman -S docker
resolving dependencies...
looking for conflicting packages...

Packages (2) bridge-utils-1.6-2  docker-1:18.02.0-1

Total Download Size:    22.99 MiB
Total Installed Size:  116.55 MiB

:: Proceed with installation? [Y/n]
```
Docker runs into issues after being successfully installed.  The fact that docker does run on device, but does not work as intended is posted for your consideration.  This might be an available architecture issue.

**Comment by T-vK** [comment]
This doesn't change anything. The PRoot would still use the kernel of your device. The only way around it would be virtualization which would probably make everything painfully slow and I'm not even sure if this can be done on any phone.

**Comment by smiller171** [comment]
Yeah, the only way to do this is to build a custom kernel for your phone unless Google decides to build container support into Android.

**Comment by T-vK** [comment]
And I doubt the latter will happen. There has been something in the news recently about Chrome OS potentially adding some kind of container support, though..

**Comment by Serkan-devel** [comment]
Or it will be implemented on Fuchisa

**Comment by smiller171** [comment]
@T-vK It actually might happen. Google is heavily invested in containers so I could see a future version of Android using container sandboxing as an alternative to the current JVM sandboxing. They wouldn't expose it as Docker natively, but it could open a path to compiling docker-engine for stuff like Termux.

**Comment by smiller171** [comment]
Also, you're thinking of this: https://chromeunboxed.com/news/chrome-os-chromebook-containers-developer which seems to be designed for improving Chromebooks for development. I'm hoping for an announcement at I/O

**Comment by emahuni** [comment]
Ok here is a guy who ran a docker image on termux. It was a 
https://www.reddit.com/r/docker/comments/7r7t6b/is_docker_possible_on_mobile/
using this: https://github.com/nmilosev/anyfed

So can't we take lessons from there and create an actual docker image running package?

**Comment by rodrigoaguilera** [comment]
I think the scope of this issue is to be able to have a docker engine running and be able to use the docker CLI client. 

Using docker images as a base for chroot is nice but I feel it does not get any closer to the goal

**Comment by T-vK** [comment]
I agree with rodrigoaguilera.

**Comment by Serkan-devel** [comment]
Maybe a seperate issue could be made for a `docker-cli`

**Comment by T-vK** [comment]
This issue has been about the real docker all along. So if a new issue should be created then it should be for `anyfed`.

**Comment by rodrigoaguilera** [comment]
I didn't try but the docker cli should run just fine from the binaries for your architecture

https://download.docker.com/linux/static/stable/
